### ISSUE TITLE
No highlighting unless file opened in fuzzy finder or file explorer

### ISSUE THREAD

**Issue by farbodsz** [nvim-treesitter/nvim-treesitter#1412]
**Describe the bug**

nvim-treesitter syntax highlighting does not always show, depending on how I open the file.

My example below shows a Haskell file, but I get this issue with all filetypes.

**To Reproduce**

When I open a file directly from the terminal, or open a new file in nvim, I get non-treesitter syntax highlighting, even though treesitter is correctly loaded and has parsed the file.

When I open it via a file explorer or fuzzy finder, I do get treesitter syntax highlighting.

The video below might show this more clearly:

https://user-images.githubusercontent.com/12140044/122764889-d7f53c80-d297-11eb-8763-3f71374a742a.mp4

**Expected behavior**

Treesitter-based highlighting when opening the file from the terminal, or creating a new file.

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found  0.19.5 (8d8690538ef0029885c7ef1f163b0e32f256a5aa) (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v14.17.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `clang` executable found. Selected from { "clang" }
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I
  - java           ✓ ✓ . ✓ ✓ 
  - kotlin         ✓ . . . ✓ 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - julia          ✓ ✓ ✓ . ✓ 
  - json           ✓ ✓ ✓ ✓ . 
  - css            ✓ . ✓ ✓ ✓ 
  - scss           ✓ . . ✓ . 
  - erlang         . . . . . 
  - elixir         ✓ ✓ . ✓ ✓ 
  - regex          ✓ . . . . 
  - ocaml          ✓ ✓ ✓ . ✓ 
  - nix            ✓ ✓ ✓ . ✓ 
  - dart           ✓ ✓ . ✓ ✓ 
  - ocaml_interface✓ ✓ ✓ . ✓ 
  - rst            ✓ ✓ . . ✓ 
  - ocamllex       ✓ . . . ✓ 
  - fennel         ✓ ✓ . . ✓ 
  - swift          . . . . . 
  - teal           ✓ ✓ ✓ ✓ ✓ 
  - c_sharp        ✓ . ✓ . ✓ 
  - ql             ✓ ✓ . ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ ✓ 
  - verilog        ✓ ✓ ✓ . ✓ 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - tsx            ✓ ✓ ✓ ✓ ✓ 
  - comment        ✓ . . . . 
  - scala          . . . . . 
  - jsdoc          ✓ . . . . 
  - supercollider  ✓ ✓ ✓ ✓ ✓ 
  - clojure        ✓ ✓ ✓ . ✓ 
  - haskell        ✓ . . . ✓ 
  - go             ✓ ✓ ✓ ✓ ✓ 
  - toml           ✓ ✓ ✓ ✓ ✓ 
  - gdscript       ✓ ✓ . . ✓ 
  - glimmer        ✓ . . . . 
  - cpp            ✓ ✓ ✓ ✓ ✓ 
  - turtle         ✓ ✓ ✓ ✓ ✓ 
  - cuda           ✓ ✓ ✓ ✓ . 
  - devicetree     ✓ ✓ ✓ ✓ ✓ 
  - dockerfile     ✓ . . . ✓ 
  - query          ✓ ✓ ✓ ✓ ✓ 
  - jsonc          ✓ ✓ ✓ ✓ ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 
  - elm            . . . . . 
  - ledger         ✓ . ✓ ✓ ✓ 
  - beancount      ✓ . ✓ . . 
  - latex          ✓ . ✓ . . 
  - dot            . . . . . 
  - gomod          ✓ . . . . 
  - bibtex         ✓ . ✓ ✓ . 
  - graphql        ✓ . . ✓ ✓ 
  - c              ✓ ✓ ✓ ✓ ✓ 
  - zig            ✓ ✓ ✓ ✓ ✓ 
  - ruby           ✓ ✓ ✓ ✓ ✓ 
  - r              ✓ ✓ . . . 
  - fortran        ✓ . ✓ ✓ . 
  - bash           ✓ ✓ ✓ . ✓ 
  - commonlisp     ✓ ✓ ✓ . . 
  - svelte         ✓ . ✓ ✓ ✓ 
  - vue            ✓ . ✓ . ✓ 
  - fish           ✓ ✓ ✓ ✓ ✓ 
  - sparql         ✓ ✓ ✓ ✓ ✓ 
  - python         ✓ ✓ ✓ ✓ ✓ 
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - php            ✓ ✓ ✓ ✓ ✓ 
  - yaml           ✓ ✓ ✓ ✓ ✓ 

 Legend: H[ighlight], L[ocals], F[olds], I[ndents]
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
</code>
</details>

**Output of `nvim --version`**

<details>
<code>
NVIM v0.5.0-dev+1429-gb585f723b
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az50-622

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/runner/work/neovim/neovim/build/nvim.AppDir/usr/share/nvim"

Run :checkhealth for more info
</code>
</details>

**Additional context**

I use `vi` as an alias for `nvim.appimage`. If it helps, my dotfiles can be found [in this repo](https://github.com/farbodsz/dotfiles).

**Comment by stsewd** [comment]
Hi, I think a plugin may be interfering with the syntax highlight, can you try with a minimal init.vim? Try disabling plugins one by one and see if the problem goes away.

**Comment by farbodsz** [comment]
Thanks @stsewd for helping me identify more details. I'm not sure on the exact cause but I can show what I found:

In [this section of my init.vim](https://github.com/farbodsz/dotfiles/blob/444d2b7df26234927d12b657870f725ffe796260/neovim/.config/nvim/init.vim#L82-L90) I source some files:

```vim
...
source ~/.config/nvim/markdown.vim
source ~/.config/nvim/syntax.vim        " <-- has my nvim-treesitter config
source ~/.config/nvim/tex.vim
```

Importantly, in `syntax.vim`, [I configure treesitter-based folds](https://github.com/farbodsz/dotfiles/blob/444d2b7df26234927d12b657870f725ffe796260/neovim/.config/nvim/syntax.vim#L33-L35).

I found that **sourcing any file after setting up folds causes the issue**.

E.g. if I source `tex.vim` before setting up folds, then it works:

```vim
...
source ~/.config/nvim/markdown.vim
source ~/.config/nvim/tex.vim
source ~/.config/nvim/syntax.vim        " <-- has my nvim-treesitter config
```

I could even put my tex config after setting up folds but by avoiding sourcing, and this also works:

```vim
...
source ~/.config/nvim/markdown.vim
source ~/.config/nvim/syntax.vim        " <-- has my nvim-treesitter config

" Tex config
let g:tex_flavor = 'latex'
let g:tex_conceal = 'abdmg'
...
```

Commenting out the lines where I setup folds also works.
